python/cuml/experimental/explainer/permutation_shap.pyx

```diff
@@ -0,0 +1,420 @@
+
+
+cdef extern from "cuml/datasets/make_permutation.hpp" namespace "ML":
+
+    void make_permutation "ML::Datasets::make_permutation"(
+        handle_t& handle,
+        float* out,
+        float* background,
+        int n_rows,
+        int n_cols,
+        float* row,
+        int* idx,
+        bool rowMajor) except +
+
+    void make_permutation "ML::Datasets::make_permutation"(
+        handle_t& handle,
+        double* out,
+        double* background,
+        int n_rows,
+        int n_cols,
+        double* row,
+        int* idx,
+        bool rowMajor) except +
+
+    void single_entry_scatter "ML::Datasets::single_entry_scatter"(
+        handle_t& handle,
+        float* out,
+        float* background,
+        int n_rows,
+        int n_cols,
+        float* row,
+        int* idx,
+        bool rowMajor) except +
+
+    void single_entry_scatter "ML::Datasets::single_entry_scatter"(
+        handle_t& handle,
+        double* out,
+        double* background,
+        int n_rows,
+        int n_cols,
+        double* row,
+        int* idx,
+        bool rowMajor) except +
+
+
+def identity(x):
+    return x
+
+
+def _identity_inverse(x):
+    return x
+
+
+def logit(x):
+    return cp.log(x / (1 - x))
+
+
+def _logit_inverse(x):
+    return 1 / (1 + cp.exp(-x))
+
+
+identity.inverse = _identity_inverse
+logit.inverse = _logit_inverse
+
+
+# kernels to do a fast calculation of the values for each row
+# depending on a shuffled index
+_row_values_kernel_f32 = cp.RawKernel(r'''
+    extern "C" __global__
+    void _row_values(float* row_values, float* averaged_outs,
+                     int* idx, int n) {
+        int tid = blockDim.x * blockIdx.x + threadIdx.x;
+        if(tid < n){
+            int pos = idx[tid];
+            row_values[pos] +=
+            2 * (averaged_outs[tid + 1] - averaged_outs[tid]);
+        }
+    }
+    ''', '_row_values')
+
+_row_values_kernel_f64 = cp.RawKernel(r'''
+    extern "C" __global__
+    void _row_values(double* row_values, double* averaged_outs,
+                     int* idx, int n) {
+        int tid = blockDim.x * blockIdx.x + threadIdx.x;
+        if(tid < n){
+            int pos = idx[tid];
+            row_values[pos] +=
+            2 * (averaged_outs[tid + 1] - averaged_outs[tid]);
+        }
+    }
+    ''', '_row_values')
+
+
+class PermutationSHAP():
+    """
+
+    Initial experimental version of a GPU accelerated of SHAP's
+    permutation explainer:
+    https://github.com/slundberg/shap/blob/master/shap/explainers/_permutation.py
+
+    This method approximates the Shapley values by iterating through
+    permutations of the inputs. Quoting the SHAP library docs, it guarantees
+    local accuracy (additivity) by iterating completely through  entire
+    permutations of the features in both forward and reverse directions.
+
+    Current limitations of the GPU version (support in progress):
+
+    - Batched, both for supporting larger daasets as well as to accelerate
```

Comment:
Sort of unclear what we mean by batched here... I was confused in our live discussion too... I think you're referring to batched over features/permutations not just rows?